sky/adaptors/aws.py

```diff
@@ -29,39 +33,70 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-# lru_cache() is thread-safe and it will return the same session object
-# for different threads.
-# Reference: https://docs.python.org/3/library/functools.html#functools.lru_cache # pylint: disable=line-too-long
-@functools.lru_cache()
 @import_package
 def session():
     """Create an AWS session."""
     # Creating the session object is not thread-safe for boto3,
     # so we add a reentrant lock to synchronize the session creation.
     # Reference: https://github.com/boto/boto3/issues/1592
-    # However, the session object itself is thread-safe, so we are
-    # able to use lru_cache() to cache the session object.
-    with _session_creation_lock:
-        return boto3.session.Session()
+
+    # Retry 5 times by default for potential credential errors,
+    # mentioned in
+    # https://github.com/skypilot-org/skypilot/pull/1988
+    if hasattr(_local, 'session'):
+        return _local.session
+
+    max_attempts = 5
+    attempt = 0
+    backoff = common_utils.Backoff()
+    err = None
+    while attempt < max_attempts:
+        try:
+            with _session_creation_lock:
+                # NOTE: we need the lock here to avoid
+                # thread-safety issues when creating the session,
+                # because Python module is a shared object,
+                # and we are not sure the if code inside
+                # boto3.session.Session() is thread-safe.
+                _local.session = boto3.session.Session()
+            return _local.session
+        except (botocore_exceptions().CredentialRetrievalError,
+                botocore_exceptions().NoCredentialsError) as e:
+            time.sleep(backoff.current_backoff())
+            err = e
+            attempt += 1
+    raise err
 
 
-@functools.lru_cache()
 @import_package
-def resource(resource_name: str, **kwargs):
-    """Create an AWS resource.
+def resource(service_name: str, **kwargs):
+    """Create an AWS resource of a certain service.
+
+    It is relatively fast to create a resource from the session (<1s), so we
+    don't need to cache the resource object.
 
     Args:
         resource_name: AWS resource name (e.g., 's3').
         kwargs: Other options.
     """
-    # Need to use the resource retrieved from the per-thread session
-    # to avoid thread-safety issues (Directly creating the client
-    # with boto3.resource() is not thread-safe).
-    # Reference: https://stackoverflow.com/a/59635814
-    return session().resource(resource_name, **kwargs)
+    if not hasattr(_local, 'resource'):
+        _local.resource = {}
+
+    # Using service name and kwargs as key
+    sorted_kwargs = tuple(sorted(kwargs.items(), key=lambda x: x[0]))
```

Comment:
key=lambda x: x[0] seems omittable?

Reply:
well, `kwargs.items()` returns a list of pair of `(key: str, value: Any)`, where we only want to compare against the key, since values may not be comparable.

Reply:
Sg. Ref I found: https://github.com/python/cpython/blob/main/Lib/functools.py#L448